### DISCUSSION TITLE
install pimcore bundles automatically

### DISCUSSION THREAD

**Discussion by BjBue** [pimcore/pimcore#13592] [Q&A / Support]
Hello,
I'm trying to automatically install the bundles (datahub, e-commerce, etc.) in docker and run the associated migrations. Unfortunately, my approaches don't work.
The bundles are specified in kernel.php.
Neither the composer post-install works
- "php bin/console pimcore:bundle:install PimcoreCustomerManagementFrameworkBundle"
nor the specification in the dockerfile
RUN php bin/console pimcore:bundle:install PimcoreCustomerManagementFrameworkBundle

Does anyone have a hint how to install it automatically?

**Comment by jdreesen** [comment]
Please add some additional information about why the `pimcore:bundle:install` command doesn't work.

**Reply by jdreesen** [reply]
Oh, you want to install the bundles when building the docker images? That probably won't work at all because they may need to access the database.

**Reply by BjBue** [reply]
Hi and thank you for reply.
I found this:
https://github.com/pimcore/demo/blob/11.x/config/packages/security.yaml
There is an installation via composer in the require-block:
"require": {
    "pimcore/customer-management-framework-bundle": "^3.3.0",
.
.
.
And why do i install it in composer and after starting i have to install it again with the symfony console, and have to do the necessary migrations every time again and again?
Sorry because my bad english

**Reply by BlackbitDevs** [reply]
Composer does not install things but it fetches the plugin's source code (including a compatibility check with other libraries and your system environment). The plugin installation with `bin/console pimcore:bundle:install <BundleName>` then installs the bundle (e.g. creating database tables).
In a docker image you can only integrate the bundle's source code but you cannot ship an installed bundle. So you have to run the bundle migrations on container startup. Just as a side note: The same applies to executing `bin/console pimcore:deployment:classes-rebuild` to update your database according to class definition files.

**Comment by mathmul** [comment]
For e-commerce and datahub, I <strike>had to follow</strike> followed different steps:

### Add e-commerce bundle

1. Open /src/Kernel.php and add the following (remove default non-existing `AppBundle::class` reference)
    ```php
    public function registerBundlesToCollection(BundleCollection $collection)
    {
        if (class_exists('\\Pimcore\\Bundle\\EcommerceFrameworkBundle\\PimcoreEcommerceFrameworkBundle')) {
            $collection->addBundle(new \Pimcore\Bundle\EcommerceFrameworkBundle\PimcoreEcommerceFrameworkBundle);
        }
    }
    ```

2. Open docker container's terminal
    ```cli
    docker-compose exec php bash
    ```

3. Install e-commerce bundle
    ```cli
    ./bin/console pimcore:bundle:install PimcoreEcommerceFrameworkBundle
    ```

### Add datahub bundle

1. Open docker container's terminal
    ```cli
    docker-compose exec php bash
    ```

2. Get datahub composer package
    ```cli
    composer require pimcore/data-hub
    ```

3. Enable datahub bundle
    ```cli
    ./bin/console pimcore:bundle:enable PimcoreDataHubBundle
    ```

4. Install datahub bundle
    ```cli
    ./bin/console pimcore:bundle:install PimcoreDataHubBundle
    ```

**Comment by StefanRickli** [comment]
Use case: my setup script allows setting up a `dev` environment after `git clone` of my project, having three possible paths:

- use existing database(s) (mysql directory copied over),
- import database dump(s),
- create empty database(s).

If I choose `empty database`, I want to have all necessary bundles already installed after my setup script.

_AFAIK_ there is no way of giving the `pimcore:bundle:install` command a list of bundles and `vendor/bin/pimcore-install --install-bundles` only accepts some default values.

My current workaround is to run `pimcore:bundle:list`, grep for the installable bundle names (because after `composer install`, they directly appear as installable, since `bundles.php` is version controlled and has the necessary entries), then loop through all of them with `pimcore:bundle:install`. But this is very slow because each time the cache is cleared and rebuilt after a bundle installation.

My wish is to have at least one of these possibilities:

- `pimcore:bundle:install`: allow a list of bundles
- `vendor/bin/pimcore-install --install-bundles`: allow for other values than only the default ones
- `vendor/bin/pimcore-install`: introduce an option, something like `--install-installable-bundles` (never mind the long name 😁)

**Reply by dvesh3** [reply]
how about using `BundleSetupEvent` which allows you extend installable bundle list? https://pimcore.com/docs/platform/Pimcore/Getting_Started/Advanced_Installation_Topics/#adding-or-removing-bundles--bundle-recommendations I guess that covers the second point in your wish list...right?

**Reply by StefanRickli** [reply]
Ha! I had this suspicion that I missed something!
Thanks, will try it out.

**Reply by StefanRickli** [reply]
Using the event listener that adds all bundles from `bundles.php` to the list of required bundles, together with #15722, I was able to get the installer to bulk-install the bundles, just with `pimcore-install --install-bundles --no-interaction`. ❤️

But alas, some bundles (PimcoreDataHubBundle, PimcoreDataImporterBundle, Web2PrintToolsBundle) don't like a bulk install without clearing the cache in-between bundle installations: I was getting `[ERROR] The metadata storage is not initialized, please run the sync-metadata-storage command to fix this issue.`
So, unfortunately, I'm going back to iterating the list of installable bundles from `bin/console pimcore:bundle:list`.

**Comment by fcaps** [comment]
Was also looking into the topic of automatic installation/updating on deploy.
Sadly  the topic to install/update and deploy via CI is not well documented, somehow i wish there was a tool to make the transitions easier for me, something like helm where you get a transition from current-status to the desired-status, removing, adding or updating bundles and/or pimcore after a deploy.

Here are some results of my research:

1. **Metadata out of sync**
This error is most likely due a missing `server_version` in the `doctrine.dbal.connection` configuration. 
My solution was to **not** execute the step `write_database_config`, removing `config/local/database.yaml`  and only executing `vendor/bin/pimcore-install --no-interaction --only-steps="setup_database, mark_migrations_as_done"` when installing a new environment.
You need to adjust  your `parameters.yaml` and add the `server_version`:
```
doctrine:
    dbal:
        connections:
            default:
                user: '%env(PIMCORE_DB_USERNAME)%'
                password: '%env(PIMCORE_DB_PASSWORD)%'
                dbname: '%env(PIMCORE_DB_NAME)%'
                host: '%env(PIMCORE_DB_HOST)%'
                port: '%env(PIMCORE_DB_PORT)%'
                mapping_types:
                    enum: string
                    bit: boolean
                server_version: '%env(PIMCORE_DB_VERSION)%'
```
My server_version was  `11.3.2-MariaDB`, you can look [here](https://symfony.com/doc/current/reference/configuration/doctrine.html) for more info about getting yours


2.  **Managing bundles (install/update/remove)**
Oh boi, this was a ride.. From what i learned you can easily corrupt your database/filesystem if you are not 100% sure what a Bundle is doing.
Let's take PimcoreDataHubBundle as example, if you "enable" it (just registering it via bundles.php or in your Kernel) and then run the migrations:migrate  after deploy (like you would do it on any other symfony app), the bundle will run migrations, will be listed as installed in `pimcore:bundle:list` but has never run the `Pimcore\Bundle\DataHubBundle\Installer::install` step, thus missing 3 tables.

@dvesh3 can you give us some insights about this issue? Or how the PimcoreDataHubBundle should be managed via CI?

**Removing** a Bundle could be as easy as a `composer remove xy` for some, but for others it's complicated and it looks like this was never a prio to be supported by pimcore. Let's take the PimcoreDataHubBundle again as example (sorry for that),  to remove the Bundle you have to execute manually tasks in the correct order, since `bin/console pimcore:bundle:uninstall PimcoreDataHubBundle` is doing nothing besides flipping a bit in the settings_store, leaving all tables, data etc. untouched.

1. somehow create your own bundle-manager that can run all this tasks for you, since you cannot use the one provided by pimcore for most of this cases .
2. remove `BUNDLE_INSTALLED__Pimcore\Bundle\DataHubBundle\PimcoreDataHubBundle` in the `settings_store` table [source](https://github.com/pimcore/data-hub/blob/1.x/src/Migrations/PimcoreX/Version20210305134111.php#L45)
3. run `bin/console doctrine:migration:execute --down VersionXXXX` for every [migration](https://github.com/pimcore/data-hub/tree/1.x/src/Migrations/PimcoreX) in this bundle AND in the the right order (maybe there is a better approach)
4. remove the config `pimcore_data_hub`, usually found in src/config.yaml or similar
5. remove var/config/data_hub/*
6. `composer remove pimcore/data-hub` with all remaining usages (bundles.php/Kernel.php)
7. remember that you can also remove `pimcore/admin-ui-classic-bundle` since this is not automatically removed.
8. deploy again
9. remove the now "missing" migrations in the `migration_version` table, or via command

My personal impression:
-  you cannot execute "bin/console doctrine:migrations:migrate --no-interaction" after deploy
- adding a bundle can be difficult and needs proper testing and inspection of the bundle-Installer
- updating a bundle is also not streamlined, meaning you need to know the bundle needs. most likely running the migrations for that bundle
- downgrading a bundle... same as remove, but just calling migration --down for the specific version + 2 deploys 
- removing a bundle in most cases (bundle that is having migrations) you need 2 deploys, one deploy to run migrations --down and a second deploy to remove the code.
- updating pimcore and core-bundles, did not look into that topic, but i can imagine it will be manual work.

All in all, pimcore was not designed to be managed via a repository/ci imo, and if you are trying todo it anyways, prepare for lot's of manual releases.
